infra/Pos/Communication/Relay/Types.hs

```diff
@@ -44,10 +45,17 @@ instance Buildable PropagationMsg where
     build (DataOnlyPM conts) =
         Buildable.build conts
 
--- | Queue of InvMsges which should be propagated.
-type RelayPropagationQueue = TBQueue (Microsecond, PropagationMsg)
+data RelayContext m = RelayContext
```

Comment:
@avieth I am not proposing to stick to `IO` or `MonadIO`. I am proposing to move methods from `RelayContext` to a type class. As I wrote already it can be:
```
class MonadRelay m where
    rlyIsPropagation :: m Bool
    rlyEnqueue :: (PropagationMsg -> Maybe NodeId -> m ())
    rlyDequeue :: (m (PropagationMsg, Maybe NodeId, NodeId, Microsecond))
```

And you also need some data to implement this interface. AFAIU from `Pos.Launcher.Resource` the only necessary data is `DiscoveryContextSum`. Btw, I think this logic (which wraps `simpleRelayer` into `RelayContext`) should be somewhere in `Pos.Communication.Relay` (unless there are fundamental difficulties in doing it).

So you can define `RelayContext` as `newtype RelayContext = RelayContext DicoveryContextSum`. If later you want to support another implementation or add more data you can always expand this type.

Then you can define functions like
```
  rlyIsPropagationDefault :: (Ether.MonadReader' RelayContext m, WhateverElseConstraintsAreNeeded m) => m Bool
  rlyEnqueue :: (Ether.MonadReader' RelayContext m, WhateverElseConstraintsAreNeeded m) => PropagationMsg -> Maybe NodeId -> m ()
  rlyDequeue :: …
```
I. e. basically do what you already do in `Launcher.Resource`.

And then you can use these definitions in `MonadRelay` instance for all execution modes which need it. For example, in `Pos.WorkMode` there are instances of many type classes for `RealMode`. It is boilerplate, but there was a trade-off between boilerplate and understandability and we decided to leave this particular part simple.

P. S. Unfortunately #1018 is still not merged, because CI was broken this night :(
I will merge it as soon as CI passes. General idea will be same with those changes. Main change is that instead of `Ether.MonadReader' RelayContext m` you'll have to write `MonadReader ctx m, HasRelayContext ctx` (because we had to get rid of `ether`).